core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/login/LoginViewModel.kt

```diff
@@ -156,9 +159,26 @@ class LoginViewModel(
         }
     }
 
-    fun onUsernameChanged(username: String) {
-        _uiState.update { prev ->
-            prev.copy(username = username)
+    fun onUsernameChanged(newValue: String) {
+        // If no input or input is valid (including whitespace)
+        val lastChar = newValue.lastOrNull()
+        val isFirstChar = newValue.length == 1
+
+        val isValid = when {
```

Comment:
This is repeating the same logic that is in the use case. That's against [coding style](https://github.com/UstadMobile/UstadMobile/blob/primary/CODING-STYLE.md#dont-repeat-yourself) - "Don't repeat yourself" -important principle in software development ( see also https://en.wikipedia.org/wiki/Don%27t_repeat_yourself ).

There is now separate logic validating the username in three places now: PersonAccountEdit, Login, and the ValidateUsernameUseCase. That is very problematic : for all reasons mentioned in the Wikipedia article e.g. more likely to have a bug, if something needs changed it needs changed in multiple places. Same concept as I referenced in Meshrabiya for converting ip addresses to Int etc.